File: app/src/main/java/activity/amigosecreto/InserirDesejoActivity.java
package activity.amigosecreto;

import android.content.Intent;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.activity.EdgeToEdge;
import androidx.core.app.NavUtils;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;

import activity.amigosecreto.db.Desejo;
import activity.amigosecreto.db.DesejoDAO;

public class InserirDesejoActivity extends AppCompatActivity {
    private Desejo desejo;
    private TextInputEditText et_produto;
    private TextInputEditText et_categoria;
    private TextInputEditText et_preco_minimo;
    private TextInputEditText et_preco_maximo;
    private TextInputEditText et_lojas;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_inserir_desejo);

        // Configurar MaterialToolbar
        MaterialToolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setNavigationOnClickListener(v -> finish());

        // Ajusta padding inferior do ScrollView quando o teclado abre (EdgeToEdge + IME inset).
        // Garante que o botão Salvar permaneça visível sem fechar o teclado manualmente.
        androidx.core.widget.NestedScrollView scrollView = findViewById(R.id.scroll_inserir_desejo);
        if (scrollView != null) {
            final int padBottomBase = scrollView.getPaddingBottom();
            ViewCompat.setOnApplyWindowInsetsListener(scrollView, (v, insets) -> {
                Insets ime = insets.getInsets(WindowInsetsCompat.Type.ime());
                Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
                int bottomInset = Math.max(ime.bottom, systemBars.bottom);
                v.setPadding(v.getPaddingLeft(), v.getPaddingTop(), v.getPaddingRight(),
                        padBottomBase + bottomInset);
                return insets;
            });
        }

        et_produto = findViewById(R.id.et_produto_ins);
        et_categoria = findViewById(R.id.et_categoria_ins);
        et_preco_minimo = findViewById(R.id.et_preco_minimo_ins);
        et_preco_maximo = findViewById(R.id.et_preco_maximo_ins);
        et_lojas = findViewById(R.id.et_lojas_ins);

        MaterialButton btnSalvar = findViewById(R.id.btn_salvar_ins);
        if (btnSalvar != null) {
            btnSalvar.setOnClickListener(v -> {
                if (validar()) {
                    inserir();
                    finish();
                }
            });
        }
    }
    
    private boolean validar() {
        if (et_produto.getText().toString().trim().isEmpty()) {
            Toast.makeText(this, "Informe o nome do produto", Toast.LENGTH_SHORT).show();
            return false;
        }
        return true;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.inserir_desejo, menu);
        return true;
    }
    
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.menu_salvar) {
            if (validar()) {
                inserir();
                finish();
            }
            return true;
        } else if (id == android.R.id.home) {
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
    
    private void inserir() {
        try {
            DesejoDAO dao = new DesejoDAO(this);
            dao.open();
            desejo = new Desejo();
            desejo.setId(dao.proximoId());
            desejo.setProduto(et_produto.getText().toString().trim());
            desejo.setCategoria(et_categoria.getText().toString().trim());

            // Tratar preços - substituir vírgula por ponto para parseDouble
            String pMin = et_preco_minimo.getText().toString().trim().replace(",", ".");
            desejo.setPrecoMinimo(pMin.isEmpty() ? 0 : Double.parseDouble(pMin));

            String pMax = et_preco_maximo.getText().toString().trim().replace(",", ".");
            desejo.setPrecoMaximo(pMax.isEmpty() ? 0 : Double.parseDouble(pMax));

            desejo.setLojas(et_lojas.getText().toString().trim());
            dao.inserir(desejo);
            dao.close();
            Toast.makeText(this, "Desejo salvo com sucesso!", Toast.LENGTH_SHORT).show();
        } catch (NumberFormatException e) {
            Toast.makeText(this, "Erro: preço inválido", Toast.LENGTH_SHORT).show();
            e.printStackTrace();
        } catch (Exception e) {
            Toast.makeText(this, "Erro ao salvar desejo: " + e.getMessage(), Toast.LENGTH_LONG).show();
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/activity/amigosecreto/InserirDesejoActivity.kt
package activity.amigosecreto

import android.os.Bundle
import android.view.MenuItem
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import activity.amigosecreto.db.Desejo
import activity.amigosecreto.repository.DesejoRepository
import activity.amigosecreto.util.WindowInsetsUtils
import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputEditText

@AndroidEntryPoint
class InserirDesejoActivity : AppCompatActivity() {

    companion object {
        /** Optional Intent extra: ID of the participant this wish belongs to.
         *  When absent (e.g. launched from ListarDesejos), participanteId stays 0 (no FK). */
        const val EXTRA_PARTICIPANTE_ID = "extra_participante_id"
    }

    private lateinit var etProduto: TextInputEditText
    private lateinit var etCategoria: TextInputEditText
    private lateinit var etPrecoMinimo: TextInputEditText
    private lateinit var etPrecoMaximo: TextInputEditText
    private lateinit var etLojas: TextInputEditText
    private lateinit var btnSalvar: MaterialButton

    @Inject lateinit var repo: DesejoRepository

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_inserir_desejo)

        findViewById<MaterialToolbar>(R.id.toolbar).setNavigationOnClickListener { finish() }
        WindowInsetsUtils.applyImeBottomPadding(findViewById(R.id.scroll_inserir_desejo))

        etProduto = findViewById(R.id.et_produto_ins)
        etCategoria = findViewById(R.id.et_categoria_ins)
        etPrecoMinimo = findViewById(R.id.et_preco_minimo_ins)
        etPrecoMaximo = findViewById(R.id.et_preco_maximo_ins)
        etLojas = findViewById(R.id.et_lojas_ins)
        btnSalvar = findViewById(R.id.btn_salvar_ins)

        btnSalvar.setOnClickListener {
            if (validar()) {
                btnSalvar.isEnabled = false
                inserir()
            }
        }
    }

    override fun onCreateOptionsMenu(menu: android.view.Menu): Boolean {
        menuInflater.inflate(R.menu.inserir_desejo, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        R.id.menu_salvar -> {
            if (validar()) {
                item.isEnabled = false
                inserir()
            }
            true
        }
        android.R.id.home -> {
            finish()
            true
        }
        else -> super.onOptionsItemSelected(item)
    }

    private fun validar(): Boolean {
        if (etProduto.text.toString().trim().isEmpty()) {
            Toast.makeText(this, R.string.error_product_name_required, Toast.LENGTH_SHORT).show()
            return false
        }
        return true
    }

    private fun inserir() {
        val desejo = try {
            Desejo().apply {
                produto = etProduto.text.toString().trim()
                categoria = etCategoria.text.toString().trim()
                val pMin = etPrecoMinimo.text.toString().trim().replace(",", ".")
                precoMinimo = if (pMin.isEmpty()) 0.0 else pMin.toDouble()
                val pMax = etPrecoMaximo.text.toString().trim().replace(",", ".")
                precoMaximo = if (pMax.isEmpty()) 0.0 else pMax.toDouble()
                lojas = etLojas.text.toString().trim()
                // participanteId is optional: 0 when launched from ListarDesejos (global wish list);
                // set when launched from ParticipanteDesejosActivity via EXTRA_PARTICIPANTE_ID.
                participanteId = intent.getIntExtra(EXTRA_PARTICIPANTE_ID, 0)
            }
        } catch (e: NumberFormatException) {
            btnSalvar.isEnabled = true
            Toast.makeText(this, R.string.error_invalid_price, Toast.LENGTH_SHORT).show()
            return
        }

        lifecycleScope.launch {
            try {
                withContext(Dispatchers.IO) { repo.inserir(desejo) }
                if (!isDestroyed && !isFinishing) {
                    Toast.makeText(this@InserirDesejoActivity, R.string.toast_wish_saved, Toast.LENGTH_SHORT).show()
                    setResult(RESULT_OK)
                    finish()
                }
            } catch (e: Exception) {
                if (!isDestroyed && !isFinishing) {
                    btnSalvar.isEnabled = true
                    invalidateOptionsMenu()
                    val msg = e.message ?: getString(R.string.error_unknown)
                    Toast.makeText(this@InserirDesejoActivity, getString(R.string.error_save_wish_format, msg), Toast.LENGTH_LONG).show()
                }
            }
        }
    }
}
